ArjunaJTA/cdi/classes/com/arjuna/ats/jta/cdi/transactional/TransactionalInterceptorBase.java

```diff
@@ -97,18 +120,55 @@ private Transactional getTransactional(InvocationContext ic) {
             return transactional;
         }
 
-        // see if the target is a stereotype
-        for (Annotation annotation : interceptedBean.getBeanClass().getAnnotations()) {
+        // check existence of the stereotype on method
+        Transactional transactionalMethod = getTransactionalAnnotationRecursive(ic.getMethod().getAnnotations());
+        if(transactionalMethod != null) return transactionalMethod;
+        // stereotype added by an extension
+        AnnotatedType<?> annTypeBean = extension.getBeanToAnnotatedTypeMapping().get(interceptedBean);
+        Transactional transactionalExtension = getTransactionalAnnotationRecursive(annTypeBean.getAnnotations());
+        if(transactionalExtension != null) return transactionalExtension;
+        // stereotype taken from intercepted bean getStereotype
+        Transactional transactionalBasedOnStereotype = null;
+        for(Class<? extends Annotation> stereotype: interceptedBean.getStereotypes()) {
+            transactionalBasedOnStereotype = stereotype.getAnnotation(Transactional.class);
+            if(transactionalBasedOnStereotype != null) return transactionalBasedOnStereotype;
+        }
+        // stereotype taken from intercepted bean getStereotypeDefinition
+        Annotation stereotypeBeanAnnotation = interceptedBean.getBeanClass().getAnnotation(Stereotype.class);
+        if(stereotypeBeanAnnotation != null) {
+            Transactional transactionalBasedOnDefinedStereotype = getTransactionalAnnotationRecursive(
+                beanManager.getStereotypeDefinition(stereotypeBeanAnnotation.annotationType()));
+            if(transactionalBasedOnDefinedStereotype != null) return transactionalBasedOnDefinedStereotype;
+        }
+        // non-contextual instances
+        Transactional transactionalNonContextual = getTransactionalAnnotationRecursive(
+            interceptedBean.getClass().getAnnotations());
+        if(transactionalNonContextual != null) return transactionalNonContextual; 
+
+        throw new RuntimeException(jtaLogger.i18NLogger.get_expected_transactional_annotation());
+    }
+
+    private Transactional getTransactionalAnnotationRecursive(Annotation... annotationsOnMember) {
+        if(annotationsOnMember == null) return null;
+        Class<? extends Annotation> stereotypeAnnotation = null;
+
+        for(Annotation annotation: annotationsOnMember) {
+            if(annotation.annotationType().equals(Transactional.class)) {
+                return (Transactional) annotation;
+            }
             if (beanManager.isStereotype(annotation.annotationType())) {
-                for (Annotation stereotyped : beanManager.getStereotypeDefinition(annotation.annotationType())) {
-                    if (stereotyped.annotationType().equals(Transactional.class)) {
-                        return (Transactional) stereotyped;
-                    }
-                }
+                stereotypeAnnotation = annotation.annotationType();
```

Comment:
I think that if there is more than one stereotype on the annotation (which is allowed), then this will override the variable and you will only inspect one of them which is incorrect.